executor/instruction/links_executor.rs

```diff
@@ -162,13 +167,27 @@ impl LinksExecutor {
     ) -> Result<TupleIterator, Box<ConceptReadError>> {
         let filter = self.filter_fn.clone();
         let check = self.checker.filter_for_row(context, &row);
-        let filter_for_row: Box<LinksFilterMapFn> = Box::new(move |item| match filter(&item) {
-            Ok(true) => match check(&item) {
-                Ok(true) | Err(_) => Some(item),
+
+        let role_var = self.links.role_type().as_variable().unwrap();
+        let row_role = if self.variable_modes.get(role_var).is_some_and(|mode| matches!(VariableMode::Input, mode)) {
+            Some(row.get(role_var.as_position().unwrap()).as_type().as_role_type().clone())
+        } else {
+            None
+        };
+
+        let filter_for_row: Box<LinksFilterMapFn> = Box::new(move |item|{
+            match filter(&item) {
+                Ok(true) => match check(&item) {
+                    Ok(true) => match verify_role(&item, row_role) {
```

Comment:
We need to check if the role == existing role, if there is an existing role. I added a test for this case in the IndexExecutor case.